internal/prometheusbp/spectest/spec.go

```diff
@@ -0,0 +1,195 @@
+package spectest
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/testutil"
+
+	"github.com/reddit/baseplate.go/errorsbp"
+)
+
+const (
+	client = "client"
+	server = "server"
+)
+
+var (
+	errPrefix         = errors.New("the prefix is not at the beginning of the metric name")
+	errClientServer   = fmt.Errorf("missing %q or %q as the second part of the metric name", client, server)
+	errLength         = errors.New("metric name should have a minimum of three parts")
+	errNotFound       = errors.New("metric not found")
+	errPrometheusLint = errors.New("prometheus GatherAndLint")
+	errDiffLabels     = errors.New("labels are incorrect")
+)
+
+// ValidateSpec validates that the Prometheus metrics being exposed from baseplate.go
+// conform to the baseplate spec.
+// metricPrefix should be either "grpc", "http", or "thirft".
+// clientOrServer should be either "client" or "server".
+func ValidateSpec(tb testing.TB, metricPrefix, clientOrServer string) {
+	tb.Helper()
+
+	var batch errorsbp.Batch
+	batch.AddPrefix("validate spec", validateSpec(metricPrefix, clientOrServer))
+
+	if err := batch.Compile(); err != nil {
+		tb.Error(err)
+	}
+}
+
+func validateSpec(prefix, clientOrServer string) error {
+	var batch errorsbp.Batch
+
+	metricFam, err := prometheus.DefaultGatherer.Gather()
+	if err != nil {
+		batch.Add(err)
+		return batch.Compile()
+	}
+
+	metricsUnderTest := buildMetricNames(prefix, clientOrServer)
+
+	for _, m := range metricFam {
+		name := *m.Name
+
+		if _, ok := metricsUnderTest[name]; !ok {
+			continue
+		}
+
+		batch.Add(lintMetric(name))
+		batch.Add(validateName(name, prefix, clientOrServer))
+
+		labels := map[string]struct{}{}
+		for _, metric := range m.GetMetric() {
+			for _, label := range metric.Label {
+				labels[*label.Name] = struct{}{}
```

Comment:
nit: since `struct{}{}` is used several times inside this file, I'd prefer to define it as a global variable to improve readability:
```
var dummySetValue = struct{}{}
```
and replaced the usages with `dummySetValue`.